### PR TITLE
Send max

### PR DESCRIPTION
**Pull request by JDonadio**
Closes https://github.com/bitpay/copay/issues/4884

* Add menu to display **send max** option
* Display wallet selector in confirm view without default selection
* Select wallet to get _send max information_ (add to cache list)
* Fix feeService
* Fix paperWallet current fee value

### CODE REVIEW

**Comment by gabrielbazan7** [comment]
It's recommended to use `$timeout` with `10` when calling to ` $ionicScrollDelegate.resize() `

**Comment by gabrielbazan7** [comment]
What is `isWallet` variable for ?

File: src/js/controllers/amount.js [hunk 1]
```diff
@@ -15,14 +13,17 @@ angular.module('copayApp.controllers').controller('amountController', function($
...
-
     $scope.isWallet = data.stateParams.isWallet;
     $scope.cardId = data.stateParams.cardId;
     $scope.toAddress = data.stateParams.toAddress;
     $scope.toName = data.stateParams.toName;
     $scope.toEmail = data.stateParams.toEmail;
     $scope.showAlternativeAmount = !!$scope.cardId;
     $scope.toColor = data.stateParams.toColor;
+    $scope.network = (new bitcore.Address($scope.toAddress)).network.name;
+    $scope.wallets = profileService.getWallets({
```
Comment (gabrielbazan7) [inline src/js/controllers/amount.js]: This both variable are unused

Comment (JDonadio) [inline src/js/controllers/amount.js]: $scope.network is used:

* line 36 > profileService.getWallets({onlyComplete: true, network: **$scope.network**})
* line 99 > feeService.getCurrentFeeValue(**$scope.network**, function(err, feePerKb) {...

**Comment by JDonadio** [comment]
> What is isWallet variable for ?

I think it's used to show/hide contact card/address box style

File: src/js/controllers/amount.js (hunk 1 above)
Comment (JDonadio) [inline src/js/controllers/amount.js]: You're right, thanks!

**Comment by JDonadio** [comment]
@bitjson @vanhue IMHO this is the best approach we can use. The current mockups are not totally consistent with the send max process (and these modify the current flow), so we need to get focused on keep each step isolated as much as we can. Currently the only change in this pull request is:

1. Send tab > select the recipient
2. Amount > set the amount/_**max amount option**_
3. Confirm > select from and confirm

**Comment by cmgustavo** [comment]
Still getting `Insufficient funds for fee`

**Comment by cmgustavo** [comment]
Second time, I choose `Send max amount`, I can see all my wallets (also with 0 bits). I shouldn't see wallets without funds.

bitpay_-_secure_bitcoin_wallet

**Comment by JDonadio** [comment]
Updated

File: src/js/controllers/confirm.js [hunk 2]
```diff
@@ -1,103 +1,199 @@
...
+          sendMax: true,
+          amount: resp.amount,
+          inputs: resp.inputs,
+          fee: resp.fee,
+          feePerKb: feePerKb,
+        };
+
+        cachedSendMax[$scope.wallet.id] = $scope.sendMaxInfo;
+
+        var msg = gettextCatalog.getString("{{fee}} will be deducted for bitcoin networking fees", {
```
Comment (isocolsky) [inline src/js/controllers/confirm.js]: Add a '.' at the end.

File: src/js/controllers/confirm.js [hunk 3]
```diff
@@ -1,103 +1,199 @@
...
+          fee: txFormatService.formatAmountStr(resp.fee)
+        });
+        var warningMsg = verifyExcludedUtxos();
+
+        if (!lodash.isEmpty(warningMsg))
+          msg += '. \n' + warningMsg;
```
Comment (isocolsky) [inline src/js/controllers/confirm.js]: Remove '.' at the beginning.

File: src/js/controllers/confirm.js [hunk 4]
```diff
@@ -1,103 +1,199 @@
...
+        function verifyExcludedUtxos() {
+          var warningMsg = [];
+          if (resp.utxosBelowFee > 0) {
+            warningMsg.push(gettextCatalog.getString("A total of {{amountBelowFeeStr}} were excluded. These funds come from UTXOs smaller than the network fee provided.", {
+              amountBelowFeeStr: txFormatService.formatAmountStr(resp.amountBelowFee)
+            }));
+          }
+
+          if (resp.utxosAboveMaxSize > 0) {
+            warningMsg.push(gettextCatalog.getString("A total of {{amountAboveMaxSizeStr}} were excluded. The maximum size allowed for a transaction was exceeded", {
```
Comment (isocolsky) [inline src/js/controllers/confirm.js]: Add '.' at the end.

File: src/js/controllers/confirm.js [hunk 5]
```diff
@@ -107,24 +203,24 @@ angular.module('copayApp.controllers').controller('confirmController', function(
...
-
   $scope.showWalletSelector = function() {
+    if (!$scope.useSendMax && ($scope.insufficientFunds || $scope.noMatchingWallet)) return;
     $scope.showWallets = true;
   };
 
   $scope.onWalletSelect = function(wallet) {
-    setWallet(wallet);
+    if ($scope.useSendMax) {
+      $scope.wallet = wallet;
```
Comment (isocolsky) [inline src/js/controllers/confirm.js]: Why not use `setWallet(wallet)`?

File: src/js/controllers/confirm.js [hunk 6]
```diff
@@ -244,28 +341,29 @@ angular.module('copayApp.controllers').controller('confirmController', function(
       return setSendError(msg);
     }
 
-    outputs.push({
+    var txp = {};
+    var amount;
+    if ($scope.useSendMax) amount = parseFloat((toAmount * unitToSatoshi));
```
Comment (isocolsky) [inline src/js/controllers/confirm.js]: Watch out for rounding errors.

File: src/js/services/feeService.js [hunk 7]
```diff
@@ -15,45 +15,45 @@ angular.module('copayApp.services').factory('feeService', function($log, $stateP
...
-      var feeLevelValue = lodash.find(levels, {
+      var feeLevelValue = lodash.find(levels[network], {
         level: feeLevel
       });
-      if (!feeLevelValue || feeLevelValue.feePerKB == null)
+
+      if (!feeLevelValue || !feeLevelValue.feePerKB)
         return cb({
-          message: 'Could not get dynamic fee for level: ' + feeLevel
+          message: gettextCatalog.getString('Could not get dynamic fee for level: ') + feeLevel
```
Comment (isocolsky) [inline src/js/services/feeService.js]: It is better to include the argument inside the string so translators can do a better job.

File: src/js/controllers/confirm.js (hunk 3 above)
Comment (JDonadio) [inline src/js/controllers/confirm.js]: Done

File: src/js/controllers/confirm.js (hunk 5 above)
Comment (JDonadio) [inline src/js/controllers/confirm.js]: You're right. Done

File: src/js/controllers/confirm.js [hunk 8]
```diff
@@ -1,103 +1,200 @@
...
+
+      ongoingProcess.set('retrievingInputs', true);
+      walletService.getSendMaxInfo($scope.wallet, {
+        feePerKb: feePerKb,
+        excludeUnconfirmedUtxos: !config.spendUnconfirmed,
+        returnInputs: true,
+      }, function(err, resp) {
+        ongoingProcess.set('retrievingInputs', false);
+        if (err) {
+          $scope.insufficientFunds = true;
```
Comment (JDonadio) [inline src/js/controllers/confirm.js]: Shouldn't be there. Done

File: src/js/services/feeService.js (hunk 7 above)
Comment (JDonadio) [inline src/js/services/feeService.js]: Done

File: src/js/controllers/confirm.js (hunk 6 above)
Comment (JDonadio) [inline src/js/controllers/confirm.js]: Great catch. Done

File: src/js/controllers/preferencesFee.js [hunk 9]
```diff
@@ -1,10 +1,14 @@
 'use strict';
 
-angular.module('copayApp.controllers').controller('preferencesFeeController', function($scope, $timeout, $ionicHistory, gettextCatalog, configService, feeService, ongoingProcess) {
+angular.module('copayApp.controllers').controller('preferencesFeeController', function($scope, $timeout, $ionicHistory, gettextCatalog, configService, feeService, ongoingProcess, popupService) {
 
   ongoingProcess.set('gettingFeeLevels', true);
-  feeService.getFeeLevels(function(levels) {
+  feeService.getFeeLevels(function(err, levels) {
```
Comment (cmgustavo) [inline src/js/controllers/preferencesFee.js]: `getFeeLevels` function doesn't return `err`. Please, check the file `feeServices.js`

Comment (cmgustavo) [inline src/js/controllers/preferencesFee.js]: Ignore. I didn't see the file updated.

File: src/js/controllers/confirm.js (hunk 5 above)
Comment (JDonadio) [inline src/js/controllers/confirm.js]: Also is not getting the send max info before build the tx

**Comment by gabrielbazan7** [comment]
ACK !

**Comment by cmgustavo** [comment]
ACK! Great job!

**Pull request merged**